R/unnest-longer.R

```diff
@@ -220,6 +233,21 @@ elt_to_long <- function(x,
     )
   }
 
+  size <- vec_size(x)
+
+  if (keep_empty && size == 0L) {
+    x <- vec_init(x)
+    size <- 1L
+
+    if (indices_include) {
+      if (is.integer(index)) {
+        index <- 1L
+      } else {
+        index <- NA_character_
```

Comment:
I am slightly questioning our decision to use `NA` when there aren't any names over `""`, like what `names2()` would return

Reply:
I'm really referring to this

``` r
tibble(
  x = list(1:2, c(a = 3, b = 4))
) %>%
  unnest_longer(x)
#> # A tibble: 4 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 <NA> 
#> 2     2 <NA> 
#> 3     3 a    
#> 4     4 b
```

Reply:
Maybe we could change ^ to use `""`, and then consistently use `NA_integer_` or `NA_character_` when we promote empty vectors to size 1? So its like:

No names anywhere:
- Positional names for existing values
- `NA_integer_` if a zero length vector was promoted to size 1

At least one name somewhere:
- Elements with names use their names
- Elements without names use `""`
- `NA_character_` if a zero length vector was promoted to size 1

Reply:
Using `""` for values without names is probably more consistent in this case

``` r
tibble(
  x = list(c(a = 1, 2), c(3, 4))
) %>%
  unnest_longer(x)
#> # A tibble: 4 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 "a"  
#> 2     2 ""   
#> 3     3 <NA> 
#> 4     4 <NA>
```

Reply:
I kind of like using `""` for unnamed elements and `NA` for elements that didn't exist. But I can see how that might be slicing things too fine.

Reply:
Isn't that what I'm suggesting?

Reply:
Like this case:

``` r
library(tidyr)
tibble(x = list(c(a = 1, 2), c(3, 4), NULL)) %>%
  unnest_longer(x, keep_empty = TRUE)
#> # A tibble: 5 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 "a"  
#> 2     2 ""   
#> 3     3 <NA>  # should be ""
#> 4     4 <NA>  # should be ""
#> 5    NA <NA> # should be NA
```

<sup>Created on 2023-01-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>